src/pythia/src/methods/balance.rs

```diff
@@ -0,0 +1,153 @@
+use anyhow::{Context, Result};
+
+use ic_cdk::{export::candid::Nat, query, update};
+
+use crate::{
+    clone_with_state,
+    jobs::withdraw,
+    log,
+    types::{
+        balance::Balances, errors::PythiaError, subscription::Subscriptions, timer::Timer,
+        whitelist, withdraw::WithdrawRequests,
+    },
+    utils::{address, canister, nat, siwe, web3},
+};
+
+/// Get the PMA address
+///
+/// # Returns
+///
+/// Returns a result with the PMA address
+#[update]
+pub async fn get_pma() -> Result<String, String> {
+    crate::utils::canister::pma()
+        .await
+        .map_err(|e| format!("failed to get the PMA: {e:?}"))
+}
+
+/// Deposit amount to the PMA
+///
+/// # Arguments
+///
+/// * `tx_hash` - 256-bit hash of the transaction, for example 0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef
+/// * `chain_id` - Unique identifier of the chain, for example Ethereum Mainnet is 1
+/// * `msg` - SIWE message, For more information, refer to the [SIWE message specification](https://eips.ethereum.org/EIPS/eip-4361)
+/// * `sig` - SIWE signature, For more information, refer to the [SIWE message specification](https://eips.ethereum.org/EIPS/eip-4361)
+///
+/// # Returns
+///
+/// Returns a result that can contain an error message
+#[update]
+pub async fn deposit(
+    chain_id: Nat,
+    tx_hash: String,
+    msg: String,
+    sig: String,
+) -> Result<(), String> {
+    _deposit(chain_id, tx_hash, msg, sig)
+        .await
+        .map_err(|e| format!("failed to deposit: {e:?}"))
+}
+
+async fn _deposit(chain_id: Nat, tx_hash: String, msg: String, sig: String) -> Result<()> {
+    let address = siwe::recover(&msg, &sig)
+        .await
+        .context(PythiaError::UnableToRecoverAddress)?;
+    if !whitelist::is_whitelisted(&address) {
+        return Err(PythiaError::UserIsNotWhitelisted.into());
+    }
+    if !Balances::is_exists(&chain_id, &address)? {
+        Balances::create(&chain_id, &address).context(PythiaError::UnableToAddNewBalance)?;
+    }
+
+    let tx = web3::get_tx(&chain_id, &tx_hash)
+        .await
+        .context(PythiaError::UnableToGetTx)?;
+
+    let receiver = tx.to.context(PythiaError::TxWithoutReceiver)?;
+    let pma = canister::pma_h160()
+        .await
+        .context(PythiaError::UnableToGetPMA)?;
+    if receiver != pma {
+        return Err(PythiaError::TxWasNotSentToPma.into());
+    }
+    Balances::save_nonce(&chain_id, &address, &nat::from_u256(&tx.nonce))
+        .context(PythiaError::UnableToSaveNonce)?;
+
+    let amount = nat::from_u256(&tx.value) - clone_with_state!(tx_fee);
+    #[allow(clippy::cmp_owned)]
+    if amount <= Nat::from(0) {
```

Comment:
This type can not so easily become const, because it does not have methods that are const calls. So I would need to use once_cell that will make this simple expression too complicated